go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -582,6 +542,197 @@ func (c *Controller) generateFRRConfigurations(ra *ratypes.RouteAdvertisements)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// createNetworkVRFsSubnetsMap creates a map of VRF -> subnets from selectedNetworks
+func createNetworkVRFsSubnetsMap(targetVRF string, selectedNetworks *selectedNetworks) map[string][]string {
+	networkVRFsSubnets := make(map[string][]string)
+	if targetVRF == "auto" {
+		for networkVRF, network := range selectedNetworks.networkVRFs {
+			subnets := selectedNetworks.networkSubnets[network]
+			networkVRFsSubnets[networkVRF] = subnets
+		}
+	} else {
+		networkVRFsSubnets[targetVRF] = selectedNetworks.subnets
+	}
+	return networkVRFsSubnets
+}
+
+// checkSubnetSliceOverlaps checks for subnets overlap within a list of subnets
+func checkSubnetSliceOverlaps(subnets []*net.IPNet) ([]string, error) {
+	for i, subnet := range subnets {
+		overlaps := util.IPNetOverlaps(subnet, subnets[i+1:]...)
+		if len(overlaps) > 0 {
+			return []string{subnet.String(), overlaps[0].String()}, nil
+		}
+	}
+	return []string{}, nil
+}
+
+// checkCrossRAOverlaps checks if subnets from current RA overlap with subnets from other RAs in the same VRF
+func (c *Controller) checkCrossRAOverlaps(ra *ratypes.RouteAdvertisements, targetVRF string, currentRASubnets []string) error {
+	allRAs, err := c.raLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, otherRA := range allRAs {
+		if otherRA.Name == ra.Name {
+			continue
+		}
+
+		// Skip RAs that target different VRF and not auto
+		if otherRA.Spec.TargetVRF != targetVRF && otherRA.Spec.TargetVRF != "auto" {
+			continue
+		}
+
+		nads, err := c.getSelectedNADs(otherRA.Spec.NetworkSelectors)
+		if err != nil {
+			return err
+		}
+		if len(nads) == 0 {
+			return fmt.Errorf("%w: failed checking cross RA overlap with RouteAdvertisement %q: no networks selected", errPending, ra.Name)
+		}
+		otherRASelectedNetworks, err := c.buildSelectedNetworkInfo(nads, sets.New(otherRA.Spec.Advertisements...))
+		if err != nil {
+			// If we can't get networks from another RA, log but don't fail
+			// This could happen if the other RA has configuration issues
+			klog.V(4).Infof("Failed to get selected networks from RouteAdvertisement %q for cross overlap check with RouteAdvertisement %q: %v", otherRA.Name, ra.Name, err)
+			continue
+		}
+
+		otherVRFSubnets := createNetworkVRFsSubnetsMap(otherRA.Spec.TargetVRF, otherRASelectedNetworks)
+		otherSubnets, hasMatchingVRF := otherVRFSubnets[targetVRF]
+		if !hasMatchingVRF || len(otherSubnets) == 0 {
+			continue
+		}
+
+		// Check for overlaps between current RA subnets and other RA subnets in the same VRF
+		for _, currentSubnet := range currentRASubnets {
+			for _, otherSubnet := range otherSubnets {
+				ipNets, err := util.ParseIPNets([]string{currentSubnet, otherSubnet})
+				if err != nil {
+					return fmt.Errorf("failed to check subnet overlap between %s and %s: %w", currentSubnet, otherSubnet, err)
+				}
+				if len(util.IPNetOverlaps(ipNets[0], ipNets[1])) == 0 {
+					continue
+				}
```

Comment:
nit: I guess this could be extracted to its own method for clarity, just as we have `checkSubnetSliceOverlaps`

Reply:
moved to checkSubnetPairOverlap.